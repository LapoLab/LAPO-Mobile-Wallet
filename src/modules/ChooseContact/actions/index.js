// @flow
import { createActions } from 'redux-actions';

export { setSendAddress } from '../../Send/actions';

export const {
  getContacts, setContacts,
} = createActions({
  GET_CONTACTS: (data = '') => data,
  SET_CONTACTS: (data = '') => data,
});
