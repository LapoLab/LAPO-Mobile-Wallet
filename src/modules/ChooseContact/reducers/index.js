import { handleActions } from 'redux-actions';
import {
  setContacts,
} from '../actions';


export const initialChooseContactState = {
  data: [],
};

// reducer
const reducer = handleActions(
  {
    [setContacts](state, { payload }) {
      return {
        ...state,
        data: payload,
      };
    },
  },
  initialChooseContactState,
);

export default reducer;
