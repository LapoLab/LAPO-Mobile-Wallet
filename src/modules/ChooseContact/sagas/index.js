import getContacts from './getContacts';

export * from '../actions';

export default {
  getContacts,
};
