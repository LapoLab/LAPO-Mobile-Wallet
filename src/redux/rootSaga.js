import { all } from 'redux-saga/effects';
import commonsSagas from '../commons/sagas';
import recoveryWordsSagas from '../modules/RecoveryWords/sagas';
import sendSagas from '../modules/Send/sagas';
import paperWordsConfirm from '../modules/PaperWordsConfirm/sagas';
import dashboard from '../modules/Dashboard/sagas';
import settings from '../modules/Settings/sagas';
import paperWords from '../modules/PaperWords/sagas';
import transactionsDetails from '../modules/TransactionDetails/sagas';
import chooseContact from '../modules/ChooseContact/sagas';
import splashScreen from '../modules/SplashScreen/sagas';
import switchContacts from '../modules/SwitchContacts/sagas';
import receive from '../modules/Receive/sagas';

// eslint-disable-next-line
const getListeners = (...args) => args.reduce((acc, nextArg) => [...acc, ...Object.values(nextArg).map(func => func())], []);

export default function* rootSaga() {
  yield all(
    getListeners(
      commonsSagas,
      recoveryWordsSagas,
      sendSagas,
      paperWordsConfirm,
      dashboard,
      settings,
      paperWords,
      transactionsDetails,
      chooseContact,
      splashScreen,
      switchContacts,
      receive,
    ),
  );
}
