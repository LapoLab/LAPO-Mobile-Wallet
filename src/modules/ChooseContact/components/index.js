/**
 * @flow
 */

import React, { Component, Fragment } from 'react';
import {
  Text,
  View,
  Image,
  TouchableOpacity,
  SafeAreaView,
  ScrollView,
  BackHandler,
  Platform,
} from 'react-native';
import LinearGradient from 'react-native-linear-gradient';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as actions from '../actions';
import * as actionsCommon from '../../../commons/actions';
import SpinnerPage from '../../../commons/components/SpinnerPage';

import SearchInput from '../../../commons/components/SearchInput';
import styles from './styles';
import { backIcon } from '../../../assets';

type Props = {
  navigation: { openDrawer: Function, dispatch: Function, goBack: Function, navigate: Function },
  getContacts: Function,
  setAddress: Function,
  setModalSend: Function,
  contacts: Array<{
    givenName: string,
    familyName: string,
    emailAddresses: Array<{ email: string, label: string }>,
    recordID: string,
  }>,
};
type State = {
  search: string,
};
class ChooseContact extends Component<Props, State> {
  state = {
    search: '',
  };

  componentDidMount() {
    const { getContacts } = this.props;
    getContacts();
    BackHandler.addEventListener('hardwareBackPress', this.handleBack);
  }

  componentWillUnmount() {
    BackHandler.removeEventListener('hardwareBackPress', this.handleBack);
  }

  handleBack = async () => {
    this.handleGoToBack();
    return true;
  };

  handleGoToBack = () => {
    const {
      navigation: { goBack, navigate },
      setModalSend,
    } = this.props;
    if (Platform.OS === 'ios') {
      navigate('Dashboard');
    } else {
      goBack();
    }
    setModalSend(true);
    return true;
  };

  handleChooseContact = (address: string) => () => {
    const {
      setAddress,
      navigation: { goBack, navigate },
      setModalSend,
    } = this.props;
    setAddress(address);
    if (Platform.OS === 'ios') {
      navigate('Dashboard');
    } else {
      goBack();
    }
    setModalSend(true);
  };

  handleChange = (search) => {
    this.setState({ search });
  };

  getFilterContact = () => {
    const { contacts } = this.props;
    const { search } = this.state;
    if (search === '') {
      return contacts;
    }
    return contacts.filter(
      item => (item.givenName && item.givenName.toLowerCase().includes(search.toLowerCase()))
        || (item.familyName && item.familyName.toLowerCase().includes(search.toLowerCase())),
    );
  };

  getSortContact = (contact) => {
    if (!contact || contact.length === 0) return [];
    return contact.sort((oldRelease, newRelease) => {
      const compareName = oldRelease.givenName && oldRelease.givenName
        .toLowerCase()
        .localeCompare(newRelease.givenName ? newRelease.givenName.toLowerCase() : '');
      const compareTitle = oldRelease.familyName && oldRelease.familyName
        .toLowerCase()
        .localeCompare(newRelease.familyName ? newRelease.familyName.toLowerCase() : '');

      // $FlowFixMe
      return compareName || compareTitle;
    });
  };

  render() {
    const contacts = this.getSortContact(this.getFilterContact());

    return (
      <LinearGradient
        start={{ x: -0.1, y: -0.1 }}
        end={{ x: 1.2, y: 1.2 }}
        locations={[0.04, 0.96]}
        colors={['#4183BD', '#093B0C']}
        style={{ flex: 1 }}
      >
        <SafeAreaView style={styles.container}>
          <View style={styles.header}>
            <TouchableOpacity onPress={this.handleGoToBack}>
              <Image style={styles.backHeader} source={backIcon} />
            </TouchableOpacity>
            <Text style={styles.titleHeader}>Choose contact</Text>
          </View>

          <SearchInput handleChange={this.handleChange} />

          <View style={styles.links}>
            <ScrollView>
              {contacts
                && contacts.map(item => (
                  <Fragment key={item.recordID}>
                    <View style={styles.info}>
                      <Text style={styles.title}>
                        {item.givenName}
                        {' '}
                        {item.familyName}
                      </Text>
                      <ScrollView>
                        {item.emailAddresses
                          && item.emailAddresses.map(email => (
                            <TouchableOpacity
                              key={`${item.recordID}-${email.email}`}
                              style={styles.link}
                              onPress={this.handleChooseContact(email.email)}
                            >
                              <Text style={styles.subTitle}>{email.email}</Text>
                            </TouchableOpacity>
                          ))}
                      </ScrollView>
                    </View>

                    <View style={styles.line} />
                  </Fragment>
                ))}
            </ScrollView>
          </View>

          <SpinnerPage />
        </SafeAreaView>
      </LinearGradient>
    );
  }
}

const mapStateToProps = state => ({
  contacts: state.chooseContact.data,
});

const mapDispatchToProps = dispatch => bindActionCreators(
  {
    getContacts: actions.getContacts,
    setAddress: actions.setSendAddress,
    setModalSend: actionsCommon.setModalSend,
  },
  dispatch,
);

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ChooseContact);
