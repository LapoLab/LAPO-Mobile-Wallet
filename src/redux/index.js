import { createStore, applyMiddleware, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';
import storage from 'redux-persist/lib/storage';
import { persistStore, persistReducer } from 'redux-persist';
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2';
import rootReducer from './rootReducer';
import rootSagas from './rootSaga';

const persistConfig = {
  key: 'root',
  storage,
  stateReconciler: autoMergeLevel2,
  blacklist: [
    'modals',
    'dashboard',
    'paperWords',
    'chooseContact',
    'send',
    'transactionsDetails',
    'alert',
    'pinsRecovery',
  ],
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const sagaMiddleware = createSagaMiddleware();
// eslint-disable-next-line no-undef
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const store = createStore(
  persistedReducer,
  composeEnhancers(applyMiddleware(sagaMiddleware)),
);
sagaMiddleware.run(rootSagas);

export const persistor = persistStore(store);
