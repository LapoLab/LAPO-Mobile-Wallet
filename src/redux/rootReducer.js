import { combineReducers } from 'redux';

import wallet from '../commons/reducers/wallet';
import pins from '../commons/reducers/pins';
import pinsRecovery from '../commons/reducers/pinsRecovery';
import modals from '../commons/reducers/modals';
import settings from '../commons/reducers/settings';
import notify from '../modules/Notify/reducers';
import paperConfirm from '../modules/PaperWordsConfirm/reducer';
import dashboard from '../modules/Dashboard/reducers';
import paperWords from '../modules/PaperWords/reducers';
import chooseContact from '../modules/ChooseContact/reducers';
import send from '../modules/Send/reducers';
import transactionsDetails from '../modules/TransactionDetails/reducer';
import alert from '../modules/Alert/reducers';

export default combineReducers({
  wallet,
  pins,
  pinsRecovery,
  notify,
  modals,
  settings,
  paperConfirm,
  dashboard,
  paperWords,
  chooseContact,
  send,
  transactionsDetails,
  alert,
});
