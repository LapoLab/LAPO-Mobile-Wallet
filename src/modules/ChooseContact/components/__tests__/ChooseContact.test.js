// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { store } from '../../../../redux';
import ChooseContact from '../index';

describe('ChooseContact page', () => {
  test('renders snapshots', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <ChooseContact />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
