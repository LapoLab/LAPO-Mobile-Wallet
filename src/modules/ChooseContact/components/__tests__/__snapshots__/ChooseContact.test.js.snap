// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChooseContact page renders snapshots 1`] = `
<BVLinearGradient
  colors={
    Array [
      -12483651,
      -16172276,
    ]
  }
  endPoint={
    Object {
      "x": 1.2,
      "y": 1.2,
    }
  }
  locations={
    Array [
      0.04,
      0.96,
    ]
  }
  startPoint={
    Object {
      "x": -0.1,
      "y": -0.1,
    }
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 30,
          "justifyContent": "flex-start",
          "marginTop": 20,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/back.png",
            }
          }
          style={
            Object {
              "flex": 1,
              "marginLeft": 12,
              "resizeMode": "contain",
              "width": 12,
            }
          }
        />
      </View>
      <Text
        style={
          Object {
            "alignItems": "center",
            "color": "white",
            "flex": 1,
            "fontFamily": "Lato-Bold",
            "fontSize": 18,
            "marginRight": 24,
            "textAlign": "center",
          }
        }
      >
        Choose contact
      </Text>
    </View>
    <View
      style={
        Object {
          "marginVertical": 15,
          "paddingHorizontal": 20,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        style={
          Object {
            "borderColor": "white",
            "borderRadius": 15,
            "borderWidth": 1,
            "color": "white",
            "fontFamily": "Lato-Regular",
            "fontSize": 18,
            "height": 32,
            "paddingHorizontal": 10,
            "paddingVertical": 5,
          }
        }
        underlineColorAndroid="transparent"
      />
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "paddingBottom": 20,
        }
      }
    >
      <RCTScrollView>
        <View />
      </RCTScrollView>
    </View>
  </RCTSafeAreaView>
</BVLinearGradient>
`;
